### PR TITLE
If there are both naive and better errors available for a given

### PR DESCRIPTION
**Pull request by dburgener**
conflict, supress the naive ones.

When reporting an error, conflict resolution creates error states for the all the terminals in the grammar, which can be a lot of errors. Some of these will hopefully classify as the more helpful error messages, but for terminals that can't shift to a valid production, lalrpop won't be able to produce a good error message.  This can result in error messages that have many naive errors burying the more helpful ones.

Check what sorts of errors we will be displaying before displaying them. If all we got is naive errors, just display those.  If we got some good errors, use those instead and discard the naive errors.

### CODE REVIEW

**Comment by dburgener** [comment]
This doesn't touch all of the situations where we are displaying naive errors.  The comments repeatedly say that naive errors *should* never happen, but in practice they happen a lot.

Another situation is if the nonterminal with the conflict is the entry point for the parser.  The errors are looking for context around the nonterminal with the conflict, but they can't find anything that can shift into the conflicting nonterminal if it's actually the entry point for the parser.  I'm hoping to get that fixed as well, but I'm not yet sure what the best way to address it is.  I might write up the problem and file a bug to discuss there and see if others have thoughts.

Neither the problem in this PR, nor the problem in the above paragraph seem to address #728 (I tested this PR against it, and it doesn't change the output.  Suggesting that the above paragraph is different is just speculation on my part.)  I suspect there are more issues here.

**Comment by dburgener** [comment]
This can likely be taken a step further.  Looking a little more at #728 that does seem to be a variation on the essential problem here.

The core problem is that when lalrpop detects a conflict, it then looks at all possible lookaheads and tries to find good examples for an error message.  In some cases, such as the test added here, one or a small number of lookaheads produce good error messages, and the rest don't.  So this PR says that if we have some good message and some bad discard the bad, which helps a lot in the case where we have a relatively small number of good messages compared to bad ones.  I think that may be the case in many grammars (in many states, a minority of all possible tokens as a lookahead can produce a valid production).  However, if the grammar either fails to find any good messages, or if there are valid productions on many lookaheads, this PR doesn't go far enough and we still end up producing many errors.

#728 is an example of such a grammar.  It looks to me as though we have some conflicts where many lookaheads make valid productions and produce good errors, as well as some conflicts where no lookaheads make valid productions and so we get 1 naive error per token in the grammar.

In general, I think we probably want to do something like what this PR does - consider all possible lookaheads, but only display the best errors.  What we really need to do though is go farther in reducing the error count.  The question is how.  With good errors, we have some things we can work with - for example we may have multiple errors for the exact same example production.  We could easily enough detect that and just pick one (for example, the first one).  I don't think that in the case of good errors, the specific example matters too much.  We could also rank the good errors more than just good/bad.  For example, is it possible that we have an ambiguity and inlining error for the same conflict?  We might just discard the ambiguity and only report the inlining which may be more useful.

If all we have is naive errors, I think things get a little more tricky.  The errors are less helpful - both to the user, as well as to lalrpop internals trying to determine what to use.  They are naive precisely because lalrpop couldn't find valid productions.  So one thing to say is that lalrpop should have found valid productions and just fix that.  For example, the situation mentioned above where the conflict is directly in a symbol that's an entry point.  If we teach lalrpop to find valid examples in that case, I'm not sure we'll see any more naive errors at all.  Even with that, I think that doing something better in the naive case than spewing a number of errors equal to the number of tokens seems worthwhile.  The question is which error to display.  Just taking the first is easy, but is there any way to say "this token would be more helpful".  Probably not - or else we would have made a good error on that token?

I know, that was a lot of text.  TL;DR - I think this PR is a good first step, but once it merges, I'd like to do follow-on work to reduce the number of errors displayed even more.

File: lalrpop/src/lr1/error/test.rs [hunk 1]
```diff
@@ -180,8 +180,69 @@ VarDecl = "let";
...
+        message.emit(&mut canvas);
+        assert!(canvas
+            .to_strings()
+            .iter()
+            .map(|r| r.to_string())
+            .collect::<Vec<String>>()
+            .join("\n")
+            .contains(text));
+        assert!(calls < unique_conflicts);
+        calls += 1;
```
Comment (Pat-Lafon) [inline lalrpop/src/lr1/error/test.rs:218]: Nit: I think the slightly clearer intention is to switch the ordering of these two and using a `<=`

File: lalrpop/src/lr1/error/test.rs [hunk 2]
```diff
@@ -180,8 +180,69 @@ VarDecl = "let";
...
+) {
+    use crate::message::{Content, Message};
+    use ascii_canvas::AsciiCanvas;
+    let _tls = Tls::test();
+    let _lr1_tls = Lr1Tls::install(grammar.terminals.clone());
+    let err = build_states(&grammar, nt(pub_state)).unwrap_err();
+    let mut cx = ErrorReportingCx::new(&grammar, &err.states, &err.conflicts);
+    let conflicts = super::token_conflicts(&err.conflicts);
+    assert_eq!(conflicts.len(), unique_conflicts); // One group of conflicts
+    assert_eq!(conflicts[0].len(), terminal_count); // terminal count
```
Comment (Pat-Lafon) [inline lalrpop/src/lr1/error/test.rs:204]: Nit: Unsure how much more extensively you plan to use this in later pr's but converting this to check terminal count with the sum of all conflicts is probably more robust.

**Review by Pat-Lafon — APPROVED** [review]
Thank you for the detailed write up. I had a nit or two but given what you've mentioned, I find this pr to be a rather clear/understandable first step.

Interested to see how much error reporting can be improved.

**Comment by Pat-Lafon** [comment]
> Another situation is if the nonterminal with the conflict is the entry point for the parser. The errors are looking for context around the nonterminal with the conflict, but they can't find anything that can shift into the conflicting nonterminal if it's actually the entry point for the parser. I'm hoping to get that fixed as well, but I'm not yet sure what the best way to address it is. I might write up the problem and file a bug to discuss there and see if others have thoughts.

👀 I think this is worth exploring.

**Comment by Pat-Lafon** [comment]
> In general, I think we probably want to do something like what this PR does - consider all possible lookaheads, but only display the best errors. What we really need to do though is go farther in reducing the error count. The question is how. With good errors, we have some things we can work with - for example we may have multiple errors for the exact same example production. We could easily enough detect that and just pick one (for example, the first one). I don't think that in the case of good errors, the specific example matters too much. We could also rank the good errors more than just good/bad. For example, is it possible that we have an ambiguity and inlining error for the same conflict? We might just discard the ambiguity and only report the inlining which may be more useful.

I imagine length of the conflict is also relevant here. If I understand correctly, shorter variations are probably going to be more helpful to users.

> If all we have is naive errors, I think things get a little more tricky. The errors are less helpful - both to the user, as well as to lalrpop internals trying to determine what to use. They are naive precisely because lalrpop couldn't find valid productions. So one thing to say is that lalrpop should have found valid productions and just fix that. For example, the situation mentioned above where the conflict is directly in a symbol that's an entry point. If we teach lalrpop to find valid examples in that case, I'm not sure we'll see any more naive errors at all. Even with that, I think that doing something better in the naive case than spewing a number of errors equal to the number of tokens seems worthwhile. The question is which error to display. Just taking the first is easy, but is there any way to say "this token would be more helpful". Probably not - or else we would have made a good error on that token?

It sounds like 'Just have LALRPOP produce only valid examples' is ideal if it is possible. I was trying to search around and see if we could take any inspiration from other parser generators. I don't think I found much that was relevant(https://www.cs.cornell.edu/andru/papers/cupex/cupex.pdf for example which didn't seem as related).

**Comment by Pat-Lafon** [comment]
Is #23 related?

**Pull request merged**